In [4]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

In [4]:
iris_df.shape

(150, 4)

So, we are dealing with 150 rows and 4 columns of data. Each row represents one datapoint and each column represents a single feature associated with the data frame. So basically, there are 150 datapoints containing 4 features each.

In [5]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

As we can see, there are four(4) columns. The columns attribute tells us the name of the columns and basically nothing else. This attribute assumes importance when we want to identify the features a dataset contains.

In [6]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


From here, we get to can make a few observations:

The DataType of each column: In this dataset, all of the data is stored as 64-bit floating-point numbers.
Number of Non-Null values: Dealing with null values is an important step in data preparation. It will be dealt with later in the notebook.

In [7]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)  
count        150.000000  
mean           1.199333  
std            0.762238  
min            0.100000  
25%            0.300000  
50%            1.300000  
75%            1.800000  
max            2.500000

The output above shows the total number of data points, mean, standard deviation, minimum, lower quartile(25%), median(50%), upper quartile(75%) and the maximum value of each column.

In [10]:
iris_df.head(10)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2
5                5.4               3.9                1.7               0.4
6                4.6               3.4                1.4               0.3
7                5.0               3.4                1.5               0.2
8                4.4               2.9                1.4               0.2
9                4.9               3.1                1.5               0.1

In [9]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

None: non-float missing data
Because None comes from Python, it cannot be used in NumPy and pandas arrays that are not of data type 'object'. Remember, NumPy arrays (and the data structures in pandas) can contain only one type of data. This is what gives them their tremendous power for large-scale data and computational work, but it also limits their flexibility. Such arrays have to upcast to the “lowest common denominator,” the data type that will encompass everything in the array. When None is in the array, it means you are working with Python objects.

In [2]:
import numpy as np 
example = np.array([2,None,6,8])
example

array([2, None, 6, 8], dtype=object)

Key takeaway: Addition (and other operations) between integers and None values is undefined, which can limit what you can do with datasets that contain them.

NaN and None: null values in pandas
Even though NaN and None can behave somewhat differently, pandas is nevertheless built to handle them interchangeably. To see what we mean, consider a Series of integers:

In [13]:
int_series = pd.Series([1,3,None,6],dtype=int())
int_series


TypeError: Cannot interpret '0' as a data type

In [16]:
example3 = pd.Series([2,1,np.nan,'',None])
example3

0       2
1       1
2     NaN
3        
4    None
dtype: object

In [17]:
example3.isnull()

0    False
1    False
2     True
3    False
4     True
dtype: bool

In [18]:
example3.isnull().sum()

2

Dropping null values

In [19]:
example3 = example3.dropna()
example3

0    2
1    1
3     
dtype: object

Because DataFrames have two dimensions, they afford more options for dropping data.

In [20]:
example4 = pd.DataFrame([[1,np.nan,4],[2,5,6],[np.nan,5,9]])
example4

0    1  2
0  1.0  NaN  4
1  2.0  5.0  6
2  NaN  5.0  9

You cannot drop a single value from a DataFrame, so you have to drop full rows or columns. Depending on what you are doing, you might want to do one or the other, and so pandas gives you options for both. Because in data science, columns generally represent variables and rows represent observations, you are more likely to drop rows of data; the default setting for dropna() is to drop all rows that contain any null values:

In [21]:
example4.dropna()

0    1  2
1  2.0  5.0  6

In [22]:
example4.dropna(axis='columns')

2
0  4
1  6
2  9

Categorical Data(Non-numeric)
First let we consider non-numeric data. In datasets, we have columns with categorical data. Eg. Gender, True or False etc.

In most of these cases, we replace missing values with the mode of the column. Say, we have 100 data points and 90 have said True, 8 have said False and 2 have not filled. Then, we can will the 2 with True, considering the full column.

In [24]:
fill_with_mode = pd.DataFrame([[1,2,None],[3,6,True],[4,5,False],[5,8,True],[6,8,None]])
fill_with_mode

0  1      2
0  1  2   None
1  3  6   True
2  4  5  False
3  5  8   True
4  6  8   None

First we find thye mode before we fill the mode.

In [26]:
fill_with_mode[2].value_counts()

2
True     2
False    1
Name: count, dtype: int64

so we will replace None with True.

In [28]:
fill_with_mode[2].fillna('True', inplace=True)
fill_with_mode

0  1      2
0  1  2   True
1  3  6   True
2  4  5  False
3  5  8   True
4  6  8   True

Numeric Data
Now, coming to numeric data. Here, we have a two common ways of replacing missing values:

Replace with Median of the row
Replace with Mean of the row
We replace with Median, in case of skewed data with outliers. This is because median is robust to outliers.

When the data is normalized, we can use mean, as in that case, mean and median would be pretty close.

In [29]:
fill_with_mean = pd.DataFrame([[1,2,np.nan],[4,8,np.nan],[np.nan,6,8],[6,np.nan,9]])
fill_with_mean

0    1    2
0  1.0  2.0  NaN
1  4.0  8.0  NaN
2  NaN  6.0  8.0
3  6.0  NaN  9.0

In [30]:
np.mean(fill_with_mean[0])

3.6666666666666665

In [32]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]), inplace=True)
fill_with_mean

C:\Users\toshiba\AppData\Local\Temp\ipykernel_16280\764752490.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  fill_with_mean[0].fillna(np.mean(fill_with_mean[0]), inplace=True)


0    1    2
0  1.000000  2.0  NaN
1  4.000000  8.0  NaN
2  3.666667  6.0  8.0
3  6.000000  NaN  9.0

In [33]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  NaN  5
3  1  6.0  7
4  2  8.0  9

In [34]:
fill_with_median[1].median()

4.0

In [36]:
fill_with_median[1].fillna(fill_with_median[1].median(), inplace=True)
fill_with_median = fill_with_median.copy()  # Ensures we're working with the actual data
print(fill_with_median)

   0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  4.0  5
3  1  6.0  7
4  2  8.0  9


C:\Users\toshiba\AppData\Local\Temp\ipykernel_16280\2317323828.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  fill_with_median[1].fillna(fill_with_median[1].median(), inplace=True)


Key takeaways:

Filling in missing values should be done when either there is less data or there is a strategy to fill in the missing data.
Domain knowledge can be used to fill in missing values by approximating them.
For Categorical data, mostly, missing values are substituted with the mode of the column.
For numeric data, missing values are usually filled in with the mean(for normalized datasets) or the median of the columns.

In [38]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

You can forward-fill null values, which is to use the last valid value to fill a null

In [39]:
example5.fillna(method='ffill')

C:\Users\toshiba\AppData\Local\Temp\ipykernel_16280\3194571100.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  example5.fillna(method='ffill')


a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

You can also back-fill to propagate the next valid value backward to fill a null:

In [40]:
example5.fillna(method='bfill')

C:\Users\toshiba\AppData\Local\Temp\ipykernel_16280\2679643382.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  example5.fillna(method='bfill')


a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

Encoding Categorical Data
Machine learning models only deal with numbers and any form of numeric data. It won't be able to tell the difference between a Yes and a No, but it would be able to distinguish between 0 and 1. So, after filling in the missing values, we need to do encode the categorical data to some numeric form for the model to understand.

Encoding can be done in two ways. We will be discussing them next.

LABEL ENCODING

Label encoding is basically converting each category to a number. For example, say we have a dataset of airline passengers and there is a column containing their class among the following ['business class', 'economy class','first class']. If Label encoding is done on this, this would be transformed to [0,1,2]. Let us see an example via code. As we would be learning scikit-learn in the upcoming notebooks, we won't use it here.

In [43]:
label = pd.DataFrame([[10,'business class'],
                        [20,'first class'],
                        [30,'economy class'],
                        [40,'economy class'],
                        [50,'economy class'],
                        [60,'business class']
                        ],columns=['ID','class'])
label

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class


To perform label encoding on the 1st column, we have to first describe a mapping from each class to a number, before replacing

In [45]:
class_labels = {'first class': 2,'economy class':1 ,'business class':0}
label['class'] = label['class'].replace(class_labels)
label

C:\Users\toshiba\AppData\Local\Temp\ipykernel_16280\2501294766.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  label['class'] = label['class'].replace(class_labels)


ID  class
0  10      0
1  20      2
2  30      1
3  40      1
4  50      1
5  60      0

Label encoding is used in either or both of the following cases :

When the number of categories is large
When the categories are in order.

ONE HOT ENCODING

Another type of encoding is One Hot Encoding. In this type of encoding, each category of the column gets added as a separate column and each datapoint will get a 0 or a 1 based on whether it contains that category. So, if there are n different categories, n columns will be appended to the dataframe.

For example, let us take the same aeroplane class example. The categories were: ['business class', 'economy class','first class'] . So, if we perform one hot encoding, the following three columns will be added to the dataset: ['class_business class','class_economy class','class_first class'].

In [46]:
one_hot = pd.DataFrame([[10,'business class'],
                        [20,'first class'],
                        [30,'economy class'],
                        [40,'economy class'],
                        [50,'economy class'],
                        [60,'business class']
                        ],columns=['ID','class'])
one_hot

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

one hot encoding on the 1st column

In [47]:
one_hot_data = pd.get_dummies(one_hot, columns = ['class'])

In [48]:
one_hot_data

ID  class_business class  class_economy class  class_first class
0  10                  True                False              False
1  20                 False                False               True
2  30                 False                 True              False
3  40                 False                 True              False
4  50                 False                 True              False
5  60                  True                False              False

When do we use one hot encoding? One hot encoding is used in either or both of the following cases :

When the number of categories and the size of the dataset is smaller.
When the categories follow no particular order.
Key Takeaways:

Encoding is done to convert non-numeric data to numeric data.
There are two types of encoding: Label encoding and One Hot encoding, both of which can be performed based on the demands of the dataset.

Removing duplicate data


In addition to missing data, you will often encounter duplicated data in real-world datasets. Fortunately, pandas provides an easy means of detecting and removing duplicate entries.

Identifying duplicates: duplicated
You can easily spot duplicate values using the duplicated method in pandas, which returns a Boolean mask indicating whether an entry in a DataFrame is a duplicate of an earlier one. Let's create another example DataFrame to see this in action.

In [50]:
example6 = pd.DataFrame({'letters' :['A','B'] * 2 + ['B']
                         ,'numbers':[1,2,1,3,3]})
example6

letters  numbers
0       A        1
1       B        2
2       A        1
3       B        3
4       B        3

In [51]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

Dropping duplicates: drop_duplicates
drop_duplicates simply returns a copy of the data for which all of the duplicated values are False:

In [52]:
example6.drop_duplicates()

letters  numbers
0       A        1
1       B        2
3       B        3

Both duplicated and drop_duplicates default to consider all columns but you can specify that they examine only a subset of columns in your DataFrame

In [54]:
example6.drop_duplicates(['letters'])

letters  numbers
0       A        1
1       B        2